### PR TITLE
grt: Correctly round up when subtracting obstructions

### PR DESCRIPTION
**Pull request by antonblanchard**
Signed-off-by: Anton Blanchard <anton@linux.ibm.com>

### CODE REVIEW

**Review by maliberty — APPROVED** [review]

**Comment by maliberty** [comment]
Btw, it is handy to put "Fixes #<issue_number>" in the commit (assuming the issue and the commit are in the same repo).  Then when the issue is merged to master the issue is automatically closed.

**Comment by antonblanchard** [comment]
Quite a lot of test case failures which I'll look into tomorrow morning.

**Comment by antonblanchard** [comment]
Quite a few tests had to be updated but this is passing the CI now.

File: src/grt/test/gcd.ok [hunk 1]
```diff
@@ -64,8 +64,10 @@ metal8            4512             0            0.00%             0 /  0 /  0
...
 ---------------------------------------------------------------------------------------
-Total           149816          3343            2.23%             0 /  0 /  0
+Total           148811          3343            2.25%             0 /  0 /  0
 
 [INFO GRT-0018] Total wirelength: 10258 um
 [INFO GRT-0014] Routed nets: 563
-No differences found.
+Differences found at line 914.
+105000 67200 113400 71400 metal1
+105000 67200 109200 71400 metal1
```
Comment (maliberty) [inline src/grt/test/gcd.ok:73]: There should be no differences in the .ok files

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/grt/test/gcd.ok (hunk 1 above)
Comment (antonblanchard) [reply src/grt/test/gcd.ok:73]: Ouch, let me fix that

**Comment by antonblanchard** [comment]
I think all the test case failures are fixed now

File: src/grt/src/fastroute/src/FastRoute.cpp [hunk 2]
```diff
@@ -533,7 +533,8 @@ void FastRouteCore::initBlockedIntervals(std::vector<int>& track_space)
     if (edge_cap > 0) {
       int reduce = 0;
       for (auto interval_it : intervals) {
-        reduce += ceil(std::abs(interval_it.upper() - interval_it.lower())
+        reduce += ceil(static_cast<float>(
+                           std::abs(interval_it.upper() - interval_it.lower()))
                        / track_space[layer - 1]);
```
Comment (maliberty) [inline src/grt/src/fastroute/src/FastRoute.cpp:538]: @luis201420 why is "track_space[layer - 1]" the divisor and not "track_space[layer]"?  It looks like we are using the wrong layer.

Comment (maliberty) [reply src/grt/src/fastroute/src/FastRoute.cpp:538]: Never mind, I got tripped up by dbTechLayer::getRoutingLevel() being 1-based which is where the -1 comes from.

File: src/grt/test/ndr_1w_3s.ok [hunk 3]
```diff
@@ -31,31 +31,30 @@
...
-met1              5439           122            2.24%             1 /  0 /  2
+met1              5388           120            2.23%             0 /  0 /  0
 met2              6475            57            0.88%             0 /  0 /  0
-met3              6142            42            0.68%             0 /  0 /  0
+met3              6142            44            0.72%             0 /  0 /  0
 met4              4625            12            0.26%             0 /  0 /  0
 met5              1480             0            0.00%             0 /  0 /  0
 ---------------------------------------------------------------------------------------
-Total            24161           233            0.96%             1 /  0 /  2
+Total            24110           233            0.97%             0 /  0 /  0
```
Comment (maliberty) [inline src/grt/test/ndr_1w_3s.ok:57]: nice

**Review by maliberty — APPROVED** [review]

**Comment by QuantamHD** [comment]
@maliberty should we merge?

**Comment by maliberty** [comment]
Re-running secure CI shows that sky130hd/aes is still failing.  Anton is taking a look.

**Comment by antonblanchard** [comment]
Here is a test case from the sky130hd aes fail: [grt-diverge-2.tar.gz](https://github.com/The-OpenROAD-Project/OpenROAD/files/10297669/grt-diverge-2.tar.gz)

With the patch in this PR applied we fail with:

```
[ERROR GPL-0307] RePlAce divergence detected. Re-run with a smaller max_phi_cof value.
```

Working with Matt, it seems that gpl routability is continually overestimating the congestion.

**Comment by antonblanchard** [comment]
As an experiment, I disabled divergence detection in gpl and ran with routability off and on:

routability disabled:
grt-no-routability

routability enabled:
grt-routability

So we might be overestimating the routing congestion, but it looks we ended up with a reasonable solution. I wonder if we should make the divergence detection less sensitive.

We might be circling around at least three issues:

- gpl routability mode is overestimating congestion
- The sky130hd PDN is too dense, especially on met5
- divergence detection is too sensitive and errors out even when we would have made it to a reasonable solution

**Comment by antonblanchard** [comment]
This is now passing sky130hd aes as of: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/commit/a947b24256d894aa12cb61f8e5267821c695c711

Likely it's just luck, but it would be great if we could work the other issues in fresh PRs.

**Comment by maliberty** [comment]
I tried with your latest version and I see in uW: [ERROR] finish__timing__setup__ws fail test: -3.68 >= -1.32
along with some more minor max cap/slew changes.  Is this ok with you?

**Comment by maliberty** [comment]
@vvbandeira I see some secure CI failures as well but they seem infrastructure related.  Would you take over helping Anton get this PR through next week.

**Comment by antonblanchard** [comment]
> I tried with your latest version and I see in uW: [ERROR] finish__timing__setup__ws fail test: -3.68 >= -1.32 along with some more minor max cap/slew changes. Is this ok with you?

Ok with me. met4 and met5 have a lot of signal routing on the microwatt design and this fix in tandem with the very dense PDN is going to significantly reduce the routing resources available. I'm hoping once we agree on PDN settings that things will improve again.

**Comment by antonblanchard** [comment]
I had to rebase due to conflicts with a recent merge. No other changes.

**Pull request merged**